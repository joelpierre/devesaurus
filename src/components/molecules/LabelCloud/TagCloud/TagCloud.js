import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { StaticQuery, graphql } from 'gatsby';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styles from '../LabelCloud.module.scss';
import Label from '../../../atoms/Label/Label';
import { mapTaxonomyIcon, mapTaxonomyTheme } from '../../../../utils';

export const PureTagCloudList = ({ allWordpressTag }) => (
  <>
    {allWordpressTag.edges.map(({ node }) => (
      <li key={node.id} className={classNames(styles['label-cloud__item'])}>
        <Label
          classes={styles['label-cloud__label']}
          link={'/tag/' + node.slug}
          theme={mapTaxonomyTheme(node.slug)}
        >
          <FontAwesomeIcon
            icon={mapTaxonomyIcon(node.slug)}
            className={styles['label-cloud__icon']}
          />
          {node.name}
        </Label>
      </li>
    ))}
  </>
);

const TagCloud = ({ classes }) => (
  <ul
    data-test="component-label-cloud"
    className={classNames(styles['label-cloud'], classes)}
  >
    <StaticQuery
      data-test="component-label-cloud-query"
      query={query}
      render={props => <PureTagCloudList {...props} />}
    />
  </ul>
);

TagCloud.defaultProps = {
  classes: undefined,
};

TagCloud.propTypes = {
  classes: PropTypes.string,
};

export default TagCloud;

const query = graphql`
  {
    allWordpressTag {
      edges {
        node {
          id
          name
          slug
          taxonomy
        }
      }
    }
  }
`;
