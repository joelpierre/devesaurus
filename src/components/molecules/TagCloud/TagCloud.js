import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { StaticQuery, graphql } from 'gatsby';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styles from './TagCloud.module.scss';
import Label from '../../atoms/Label/Label';
import { mapTaxonomyIcon, mapTaxonomyTheme } from '../../../utils';

export const getCloudQuery = taxonomy => {
  let query;

  switch (taxonomy) {
    case 'post_tag':
      query = tagQuery;
      return query;

    case 'category':
      query = categoryQuery;
      return query;

    case 'word_tag':
      query = wordTagQuery;
      return query;

    case 'word_category':
      query = wordCategoryQuery;
      return query;

    default:
      query = wordTagQuery;
      return query;
  }
};

export const PureTagCloud = ({ classes, taxonomy, ...props }) => {
  let obj;
  let link;

  switch (taxonomy) {
    case 'post_tag':
      obj = props.allWordpressTag;
      link = `/tag/`;
      break;

    case 'category':
      obj = props.allWordpressCategory;
      link = `/category/`;
      break;

    case 'word_tag':
      obj = props.allWordpressWpWordTag;
      link = `/word-tag/`;
      break;

    case 'word_category':
      obj = props.allWordpressWpWordCategory;
      link = `/word-category/`;
      break;

    default:
      obj = props.allWordpressWpWordTag;
      link = `/word-tag/`;
  }

  return (
    <ul
      data-test="component-tag-cloud"
      className={classNames(styles['tag-cloud'], classes)}
    >
      {obj &&
        obj.edges.map(({ node }) => (
          <li key={node.id} className={classNames(styles['tag-cloud__item'])}>
            <Label
              classes={styles['tag-cloud__label']}
              link={link + node.slug}
              theme={mapTaxonomyTheme(node.slug)}
            >
              <FontAwesomeIcon
                icon={mapTaxonomyIcon(node.slug)}
                className={styles['tag-cloud__icon']}
              />
              {node.name}
            </Label>
          </li>
        ))}
    </ul>
  );
};

const TagCloud = ({ classes, taxonomy }) => {
  return (
    <StaticQuery
      data-test="component-tag-cloud-query"
      query={getCloudQuery(taxonomy)}
      render={props => (
        <PureTagCloud {...props} taxonomy={taxonomy} classes={classes} />
      )}
    />
  );
};

TagCloud.defaultProps = {
  classes: undefined,
  taxonomy: 'word_tag',
};

TagCloud.propTypes = {
  classes: PropTypes.string,
  taxonomy: PropTypes.oneOf(['category', 'tag', 'word_tag', 'word_category']),
};

export default TagCloud;

const wordTagQuery = graphql`
  {
    allWordpressWpWordTag {
      edges {
        node {
          id
          name
          slug
          taxonomy
        }
      }
    }
  }
`;

const tagQuery = graphql`
  {
    allWordpressTag {
      edges {
        node {
          id
          name
          slug
          taxonomy
        }
      }
    }
  }
`;

const categoryQuery = graphql`
  {
    allWordpressCategory {
      edges {
        node {
          id
          name
          slug
          taxonomy
        }
      }
    }
  }
`;

const wordCategoryQuery = graphql`
  {
    allWordpressWpWordCategory {
      edges {
        node {
          id
          name
          slug
          taxonomy
        }
      }
    }
  }
`;
