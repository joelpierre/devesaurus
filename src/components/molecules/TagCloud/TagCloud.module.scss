@import "~sass/core";

.tag-cloud {
  display: flex;
}
