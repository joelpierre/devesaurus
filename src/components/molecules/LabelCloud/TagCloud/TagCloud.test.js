import React from 'react';
import { shallow } from 'enzyme/build';

import TagCloud, { PureTagCloudList } from './TagCloud';
import {
  checkProps,
  findByTestAttr,
  matchSnapshot,
} from '../../../../utils/test';

const defaultProps = {};
const defaultPureProps = {
  allWordpressTag: {
    edges: [
      {
        node: {
          id: 1,
          slug: 'test-slug',
          name: 'test-name',
          taxonomy: 'test-tax',
        },
      },
    ],
  },
};

/**
 * Factory function to create a ShallowWrapper for the TagCloud component.
 * @param {object} props - Component props specific to setup
 * @returns {ShallowWrapper}
 */
const setup = (props = {}) => {
  const setupProps = { ...defaultPureProps, ...props };
  return shallow(<TagCloud {...setupProps} />);
};
/**
 * Factory function to create a ShallowWrapper for the PureTagCloud component.
 * @param {object} props - Component props specific to setup
 * @returns {ShallowWrapper}
 */
const pureSetup = (props = {}) => {
  const setupProps = { ...defaultPureProps, ...props };
  return shallow(<PureTagCloudList {...setupProps} />);
};

describe('<PureTagCloudList/>', () => {
  let wrapper;
  let pureWrapper;

  beforeEach(() => {
    wrapper = setup();
    pureWrapper = pureSetup();
  });

  it('renders the TagCloudList Component without errors', () => {
    const component = findByTestAttr(wrapper, 'component-tag-cloud');
    expect(component.length).toBe(1);
  });

  it('renders the TagCloudList Query without errors', () => {
    const component = findByTestAttr(wrapper, 'component-tag-cloud-query');
    expect(component.length).toBe(1);
  });

  it('renders the PureTagCloudList Component without errors', () => {
    const component = findByTestAttr(pureWrapper, 'tag-cloud-item');
    expect(component.length).toBe(1);
  });

  it('should render with correct props', () => {
    checkProps(PureTagCloudList, defaultProps);
  });

  it('should match snapshot PureTagCloudList', () => {
    matchSnapshot(pureWrapper);
  });

  it('should match snapshot TagCloudList', () => {
    matchSnapshot(wrapper);
  });
});
