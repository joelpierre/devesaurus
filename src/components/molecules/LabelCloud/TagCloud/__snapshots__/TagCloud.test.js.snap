// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PureTagCloudList/> should match snapshot PureTagCloudList 1`] = `
<Fragment>
  <li
    className="label-cloud__item"
    data-test="tag-cloud-item"
    key="1"
  >
    <Label
      caps={false}
      classes="label-cloud__label"
      link="/tag/test-slug"
      size="sm"
      theme="brand"
    >
      <FontAwesomeIcon
        border={false}
        className="label-cloud__icon"
        fixedWidth={false}
        flip={null}
        icon={
          Array [
            "far",
            "atom-alt",
          ]
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
      test-name
    </Label>
  </li>
</Fragment>
`;

exports[`<PureTagCloudList/> should match snapshot TagCloudList 1`] = `
<ul
  className="label-cloud"
  data-test="component-tag-cloud"
>
  <mockConstructor
    data-test="component-tag-cloud-query"
    render={[Function]}
  />
</ul>
`;
